File: crates/daemon/sidecar/protocol.ts
/**
 * The flycod⇄sidecar protocol, mirrored from
 * `crates/daemon/src/harness/claude/protocol.rs`.
 *
 * One JSON object per line in each direction. Every message is internally
 * tagged on `type` and uses `snake_case` field names — including where the
 * Agent SDK's own types are `camelCase`, which `sidecar.ts` translates at
 * the boundary so this protocol has exactly one naming convention.
 *
 * These are zod schemas rather than bare interfaces because the sidecar
 * validates every inbound line: a command it cannot decode is a contract
 * break between two halves of the same program, and it fails loudly instead
 * of guessing. `fixtures/protocol/` pins these declarations against the Rust
 * ones — see `protocol.test.ts`.
 */
import { z } from "zod";

/**
 * Any JSON value, required to be present.
 *
 * `z.unknown()` would also accept a missing key, which would let a field
 * silently disappear from the protocol; this keeps presence part of the
 * contract while passing the value through untouched.
 */
const jsonValue = z.custom<unknown>((value) => value !== undefined, {
  message: "expected a JSON value to be present",
});

/** Mirrors the SDK's own `PermissionMode` union. */
export const permissionModeSchema = z.enum([
  "default",
  "acceptEdits",
  "bypassPermissions",
  "plan",
  "dontAsk",
  "auto",
]);

/** How the sidecar authenticates the `claude` CLI it supervises. */
export const sidecarAuthSchema = z.discriminatedUnion("mode", [
  z.object({ mode: z.literal("inherit") }),
  z.object({ mode: z.literal("oauth_token"), token: z.string() }),
  z.object({ mode: z.literal("api_key"), key: z.string() }),
]);

/**
 * One MCP server, in Claude Code's own vocabulary.
 *
 * The single place this protocol is not `snake_case`, and for the same
 * reason `sdk_message` is opaque: the value is the SDK's own
 * `McpServerConfig`, not a flycod restatement of it, so it travels from
 * `crates/daemon/src/mount.rs` into `Options.mcpServers` untouched. It is
 * also, byte for byte, what the machine's root-owned `managed-mcp.json`
 * declares — one set of servers, described once.
 *
 * Nothing here is optional even where the SDK's own type allows it: under
 * `exactOptionalPropertyTypes` a `string[] | undefined` is not a `string[]`,
 * and requiring the empty value keeps one spelling for one server on both
 * sides of the protocol.
 */
export const claudeMcpServerSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("stdio"),
    command: z.string(),
    args: z.array(z.string()),
    env: z.record(z.string(), z.string()),
    alwaysLoad: z.boolean(),
  }),
  z.object({
    type: z.literal("http"),
    url: z.string(),
    headers: z.record(z.string(), z.string()),
  }),
]);

/**
 * Where one mounted server has got to.
 *
 * `pending` is its own state rather than "not connected": a CLI still
 * dialling a server has not answered the question yet, and flycod waits it
 * out instead of refusing the session for being asked early.
 */
export const mountStateSchema = z.enum(["connected", "pending", "failed"]);

/** What the CLI reports about one server it was told to mount. */
export const mountedServerSchema = z.object({
  name: z.string(),
  status: z.string(),
  state: mountStateSchema,
  tools: z.array(z.string()),
});

/**
 * One model the harness offers, in flyco's own vocabulary.
 *
 * `snake_case`, like the rest of this protocol, and deliberately not the
 * SDK's `ModelInfo`: the same shape has to come back from Codex, so the
 * translation happens here at the boundary rather than in three readers
 * downstream. `efforts` is empty for a model that accepts none, which is a
 * real answer rather than a missing one.
 */
export const modelOptionSchema = z.object({
  id: z.string(),
  label: z.string(),
  description: z.string(),
  is_default: z.boolean(),
  efforts: z.array(z.string()),
  default_effort: z.string().nullable(),
});

/**
 * One rolling plan window, as this sidecar read it from the SDK.
 *
 * The reading and not the finished window: the *label* ("5-hour",
 * "Weekly (Fable)") is derived in Rust, because the same rule has to serve
 * Codex and writing it twice would be two vocabularies. What the sidecar
 * contributes is what only it knows — that the SDK's `five_hour` key means
 * three hundred minutes, and that a `model_scoped` row is a weekly window
 * scoped to one model.
 *
 * `used_percent` is an integer 0-100: the SDK's `utilization` is an
 * unbounded number, and it is rounded and clamped here so the wire carries
 * one spelling of a percentage.
 */
export const usageWindowSchema = z.object({
  window_minutes: z.number().int().nullable(),
  scope: z.string().nullable(),
  used_percent: z.number().int().min(0).max(100),
  resets_at_unix: z.number().int().nullable(),
});

/**
 * One slash command the running CLI offers, in flyco's own vocabulary.
 *
 * `snake_case` and not the SDK's `SlashCommand`, for the reason the model
 * option above is not `ModelInfo`: the same shape has to come back from
 * Codex. `argument_hint` is `null` for a command that takes no argument —
 * the SDK spells that as an empty string, and a palette that has to tell
 * `""` from "unset" is a palette with a bug waiting in it. Aliases are
 * dropped: `/cost` and `/stats` are two more rows saying what `/usage`
 * already says.
 */
export const harnessCommandSchema = z.object({
  name: z.string(),
  description: z.string(),
  argument_hint: z.string().nullable(),
});

/** The SDK's `SessionKey`, in this protocol's `snake_case`. */
export const sessionKeySchema = z.object({
  project_key: z.string(),
  session_id: z.string(),
  subpath: z.string().optional(),
});

/**
 * One thing occupying the context window, as a context panel lists it.
 *
 * One shape for every section the CLI reports — a usage category, an MCP
 * tool's schema, a memory file, an agent, a skill's frontmatter — because
 * each answers the same question (what is this, and what does it cost) and
 * the panel's rows differ only in which list they came from. `deferred` is
 * counted toward the window but not materialized in it yet — a deferred
 * MCP tool's schema, for instance, summarized until first called.
 */
export const contextCostSchema = z.object({
  name: z.string(),
  tokens: z.number().int().nonnegative(),
  deferred: z.boolean(),
});

/** How much of a model's context window is spoken for. */
export const contextWindowSchema = z.object({
  used_tokens: z.number().int().nonnegative(),
  size_tokens: z.number().int().nonnegative(),
});

/**
 * What the context window is spent on — the answer to `/context`.
 *
 * The SDK's `getContextUsage` answers in its own shape (`totalTokens`,
 * `maxTokens`, `isDeferred`, …) and the mapping lives in the sidecar for
 * the same reason `usageWindowSchema` does: this shape has to come back
 * from Codex spelled the same way, so the vocabulary is flyco's from here.
 * `window.size_tokens` is the SDK's `maxTokens` — the effective ceiling
 * its `percentage` is measured against, not the model's `rawMaxTokens`.
 */
export const contextUsageSchema = z.object({
  model: z.string().optional(),
  window: contextWindowSchema.optional(),
  auto_compact: z.number().int().nonnegative().optional(),
  categories: z.array(contextCostSchema),
  mcp_tools: z.array(contextCostSchema),
  memory_files: z.array(contextCostSchema),
  agents: z.array(contextCostSchema),
  skills: z.array(contextCostSchema),
});

/** One `SessionStore` operation flycod must perform. */
export const storeOpSchema = z.union([
  z.object({
    append: z.object({ key: sessionKeySchema, entries: z.array(jsonValue) }),
  }),
  z.object({ load: z.object({ key: sessionKeySchema }) }),
]);

/** A command flycod writes to the sidecar. */
export const sidecarCommandSchema = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("start"),
    cwd: z.string(),
    auth: sidecarAuthSchema,
    config_dir: z.string().nullable(),
    project_dir_name: z.string().nullable(),
    model: z.string().nullable(),
    effort: z.string().nullable(),
    permission_mode: permissionModeSchema,
    resume_session_id: z.string().nullable(),
    mcp_servers: z.record(z.string(), claudeMcpServerSchema),
    strict_mcp_config: z.boolean(),
  }),
  z.object({ type: z.literal("user_message"), text: z.string() }),
  z.object({ type: z.literal("interrupt") }),
  z.object({ type: z.literal("compact") }),
  z.object({ type: z.literal("context_usage") }),
  z.object({
    type: z.literal("set_model"),
    model: z.string(),
    effort: z.string().nullable(),
  }),
  z.object({
    type: z.literal("set_permission_mode"),
    mode: permissionModeSchema,
  }),
  z.object({
    type: z.literal("approval_decision"),
    id: z.string(),
    allow: z.boolean(),
    updated_input: jsonValue,
    message: z.string().nullable(),
  }),
  z.object({
    type: z.literal("store_response"),
    id: z.number().int(),
    result: jsonValue,
  }),
  z.object({ type: z.literal("shutdown") }),
]);

/** An event the sidecar writes to flycod. */
export const sidecarEventSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("ready"), sdk_version: z.string() }),
  z.object({ type: z.literal("started"), session_id: z.string() }),
  z.object({ type: z.literal("capabilities"), capabilities: z.array(z.string()) }),
  z.object({ type: z.literal("models"), models: z.array(modelOptionSchema) }),
  z.object({ type: z.literal("plan_usage"), windows: z.array(usageWindowSchema) }),
  z.object({ type: z.literal("context_usage"), usage: contextUsageSchema }),
  z.object({ type: z.literal("commands"), commands: z.array(harnessCommandSchema) }),
  z.object({ type: z.literal("mcp_servers"), servers: z.array(mountedServerSchema) }),
  z.object({ type: z.literal("sdk_message"), message: jsonValue }),
  z.object({
    type: z.literal("approval_request"),
    id: z.string(),
    tool: z.string(),
    input: jsonValue,
    suggestions: jsonValue,
  }),
  z.object({
    type: z.literal("store_request"),
    id: z.number().int(),
    op: storeOpSchema,
  }),
  z.object({ type: z.literal("fatal"), error: z.string() }),
]);

/** The permission mode the session runs under. */
export type PermissionMode = z.infer<typeof permissionModeSchema>;
/** How the supervised CLI authenticates. */
export type SidecarAuth = z.infer<typeof sidecarAuthSchema>;
/** One MCP server, as the SDK's `mcpServers` option takes it. */
export type ClaudeMcpServer = z.infer<typeof claudeMcpServerSchema>;
/** One model the harness offers. */
export type ModelOption = z.infer<typeof modelOptionSchema>;
/** One rolling plan window, as read from the SDK. */
export type UsageWindow = z.infer<typeof usageWindowSchema>;
/** One slash command the harness offers. */
export type HarnessCommand = z.infer<typeof harnessCommandSchema>;
/** Where one mounted MCP server has got to. */
export type MountState = z.infer<typeof mountStateSchema>;
/** What the CLI reports about one mounted MCP server. */
export type MountedServer = z.infer<typeof mountedServerSchema>;
/** The SDK's `SessionKey`, in this protocol's spelling. */
export type SessionKey = z.infer<typeof sessionKeySchema>;
/** One `SessionStore` operation. */
export type StoreOp = z.infer<typeof storeOpSchema>;
/** One thing occupying the context window. */
export type ContextCost = z.infer<typeof contextCostSchema>;
/** What the context window is spent on. */
export type ContextUsage = z.infer<typeof contextUsageSchema>;
/** A command flycod writes to the sidecar. */
export type SidecarCommand = z.infer<typeof sidecarCommandSchema>;
/** An event the sidecar writes to flycod. */
export type SidecarEvent = z.infer<typeof sidecarEventSchema>;
/** The `start` command, which opens the session. */
export type StartCommand = Extract<SidecarCommand, { type: "start" }>;

/**
 * A single-line description of a failure, suitable for a `fatal` event.
 *
 * A zod failure is rendered from its issue list rather than its default
 * JSON dump, so a protocol mismatch reads as a sentence in flycod's log.
 */
export function describeError(error: unknown): string {
  if (error instanceof z.ZodError) {
    return `the line does not match the protocol: ${z
      .prettifyError(error)
      .split("\n")
      .map((part) => part.trim())
      .filter((part) => part.length > 0)
      .join("; ")}`;
  }
  return error instanceof Error ? error.message : String(error);
}

File: crates/daemon/src/harness/claude/protocol.rs
//! The flycod⇄sidecar protocol.
//!
//! One JSON object per line in each direction over the sidecar's stdin and
//! stdout. Every message is internally tagged on `type` and uses
//! `snake_case` field names; [`sidecar/protocol.ts`] mirrors these
//! declarations as zod schemas and the fixtures under `fixtures/protocol/`
//! pin both sides to the same bytes.
//!
//! The protocol deliberately keeps the SDK's own messages opaque
//! ([`SidecarEvent::SdkMessage`] carries them verbatim): the sidecar's job
//! is transport and callback plumbing, and every interpretation of an SDK
//! message happens in Rust, in [`super::normalize`].
//!
//! [`sidecar/protocol.ts`]: https://github.com/lexoliu/flyco/blob/main/crates/daemon/sidecar/protocol.ts

use std::collections::BTreeMap;
use std::path::PathBuf;

use flyco_core::ApprovalId;
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub use crate::mount::{ClaudeMcpServer, MountedServer};

/// Identifies one outstanding [`StoreOp`] round trip.
///
/// Minted by the sidecar as a monotonic counter — unlike an
/// [`ApprovalId`], a store request never leaves the daemon, so it needs no
/// globally unique form.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(transparent)]
pub struct StoreRequestId(u64);

impl StoreRequestId {
    /// Wraps a raw counter value.
    #[must_use]
    pub const fn new(value: u64) -> Self {
        Self(value)
    }
}

/// The permission mode the Claude Agent SDK runs the session under.
///
/// Defined in [`flyco_core`] because the control plane writes it into the
/// `flycod` configuration it provisions, and re-exported here so the
/// sidecar protocol reads as one module.
pub use flyco_core::PermissionMode;

/// How the sidecar authenticates the `claude` CLI it supervises.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "mode", rename_all = "snake_case")]
pub enum SidecarAuth {
    /// Use whatever login the host user already has.
    ///
    /// No `CLAUDE_CONFIG_DIR` override and no credential injection: the CLI
    /// reads the default `~/.claude` tree. This is the developer-machine
    /// mode; a session VM never runs it.
    Inherit,
    /// Inject a Claude subscription OAuth token.
    OauthToken {
        /// Value for `CLAUDE_CODE_OAUTH_TOKEN`.
        token: String,
    },
    /// Inject an Anthropic API key.
    ApiKey {
        /// Value for `ANTHROPIC_API_KEY`.
        key: String,
    },
}

/// A command flycod writes to the sidecar.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SidecarCommand {
    /// Open the session. Sent exactly once, after [`SidecarEvent::Ready`].
    Start {
        /// Working directory the harness operates in.
        cwd: PathBuf,
        /// Credentials for the supervised CLI.
        auth: SidecarAuth,
        /// `CLAUDE_CONFIG_DIR` — an isolated config tree. `None` leaves the
        /// variable unset, which is only correct under
        /// [`SidecarAuth::Inherit`].
        config_dir: Option<PathBuf>,
        /// `CLAUDE_CODE_PROJECT_DIR_NAME` — the stable project key the SDK
        /// files sessions under.
        project_dir_name: Option<String>,
        /// Model override; `None` leaves the CLI's own default in place.
        model: Option<String>,
        /// Reasoning effort for that model, as the SDK's `EffortLevel`
        /// spells it. `None` leaves the CLI's own default for the model in
        /// place, which is the only correct value for a model that accepts
        /// no effort levels at all.
        effort: Option<String>,
        /// Permission mode for the session.
        permission_mode: PermissionMode,
        /// Harness-native session id to resume, for cross-host History.
        resume_session_id: Option<String>,
        /// Whether the CLI should be told to use only the servers below and
        /// ignore every scope it would otherwise discover.
        ///
        /// False on a provisioned machine, and it has to be: the root-owned
        /// `managed-mcp.json` there is an *enterprise* MCP config, and the
        /// CLI refuses to start at all when one is present and
        /// `--strict-mcp-config` is also asked for (issue #195). The two
        /// are the same guarantee by different means — exclusivity the
        /// agent cannot reach, and exclusivity on the command line — so
        /// exactly one of them is ever in force.
        strict_mcp_config: bool,
        /// Every MCP server this session may reach, keyed by the name the
        /// harness announces it under.
        ///
        /// Passed even on a machine whose `managed-mcp.json` already
        /// declares the same set, and the two jobs are different: the
        /// managed file makes the set *exclusive*, this makes it *present*.
        /// A developer's flycod is not root and writes no managed file, and
        /// its session still gets flyco's tools.
        mcp_servers: BTreeMap<String, ClaudeMcpServer>,
    },
    /// Push one user message into the streaming-input generator.
    UserMessage {
        /// The message text.
        text: String,
    },
    /// End the current turn through the SDK's `interrupt()` (SIGINT
    /// semantics). Never signal the child process directly.
    Interrupt,
    /// Run Claude Code's native `/compact` command.
    Compact,
    /// Put the running query on another model, at another effort.
    ///
    /// Two SDK calls behind one command, because they are one decision:
    /// `setModel` moves the query and `applyFlagSettings` sets the effort
    /// on it, and an effort applied to the model before it would be a level
    /// of something the session is no longer running.
    SetModel {
        /// The identifier the SDK's own model list names, verbatim.
        model: String,
        /// The effort level, or `None` to clear whatever was set and leave
        /// the CLI's default for the new model in place.
        effort: Option<String>,
    },
    /// Resolve a pending [`SidecarEvent::ApprovalRequest`].
    ApprovalDecision {
        /// The approval being decided.
        id: ApprovalId,
        /// Whether the tool call may proceed.
        allow: bool,
        /// Replacement tool input. Only meaningful when `allow` is true;
        /// `None` means the sidecar echoes the original input back to the
        /// SDK, which requires an `updatedInput` on every allow.
        updated_input: Option<Value>,
        /// Reason shown to the model. Required in practice when `allow` is
        /// false — the SDK's deny branch carries a message.
        message: Option<String>,
    },
    /// Resolve a pending [`SidecarEvent::StoreRequest`].
    StoreResponse {
        /// The request being answered.
        id: StoreRequestId,
        /// `null` for an append, and for a load of a stream that was never
        /// written — which is the SDK's own "no such session" signal, not
        /// an empty array. Otherwise the array of entries.
        result: Value,
    },
    /// Close the streaming-input generator and exit cleanly.
    Shutdown,
}

impl SidecarCommand {
    /// The `type` tag this command serializes under.
    #[must_use]
    pub const fn tag(&self) -> &'static str {
        match self {
            Self::Start { .. } => "start",
            Self::UserMessage { .. } => "user_message",
            Self::Interrupt => "interrupt",
            Self::Compact => "compact",
            Self::SetModel { .. } => "set_model",
            Self::ApprovalDecision { .. } => "approval_decision",
            Self::StoreResponse { .. } => "store_response",
            Self::Shutdown => "shutdown",
        }
    }
}

/// The `SessionKey` the Agent SDK's `SessionStore` addresses entries by.
///
/// The SDK spells these `projectKey` / `sessionId`; the sidecar translates
/// to this protocol's `snake_case` on the way out and back on the way in,
/// so the flycod protocol has exactly one naming convention.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct SessionKey {
    /// Stable project key, from `CLAUDE_CODE_PROJECT_DIR_NAME`.
    pub project_key: String,
    /// The SDK session this batch belongs to.
    pub session_id: String,
    /// Sub-stream within the session — subagent transcripts arrive as
    /// `subagents/agent-{id}`. Absent for the main transcript; the SDK
    /// treats an empty string as invalid.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub subpath: Option<String>,
}

/// One `SessionStore` operation the sidecar needs flycod to perform.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StoreOp {
    /// Mirror a batch of transcript entries.
    Append {
        /// Where the entries belong.
        key: SessionKey,
        /// The raw SDK entries. Most carry a `uuid`, which the store uses
        /// as an idempotency key; the SDK documents that some (titles,
        /// tags, mode markers) carry none and are appended without dedup.
        entries: Vec<Value>,
    },
    /// Read a stream back, for cross-host resume.
    Load {
        /// Which stream to read.
        key: SessionKey,
    },
}

/// One rolling plan window as the sidecar read it from the SDK.
///
/// The *reading*, not the finished [`flyco_core::UsageWindow`]: the label
/// is derived in Rust so that a five-hour window is called the same thing
/// whichever harness reported it, and a rule that lived in TypeScript would
/// have to be written a second time for Codex. So the sidecar translates
/// only what it alone knows — that the SDK's `five_hour` key means three
/// hundred minutes, and that a `model_scoped` row is a weekly window scoped
/// to one model — and hands the numbers on.
///
/// `used_percent` is an integer because that is what the protocol pins: the
/// SDK's `utilization` is an unbounded 0–100 number, and rounding it at the
/// boundary keeps one spelling of a percentage on the wire.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SidecarUsageWindow {
    /// How long the window is, in minutes.
    pub window_minutes: Option<u32>,
    /// The part of the plan this window covers, when it covers only part of
    /// one — the display name of a per-model weekly bucket.
    pub scope: Option<String>,
    /// How much of the window is spent, 0–100.
    pub used_percent: u8,
    /// When the window turns over, seconds since the Unix epoch.
    pub resets_at_unix: Option<i64>,
}

impl From<SidecarUsageWindow> for flyco_core::UsageWindow {
    fn from(window: SidecarUsageWindow) -> Self {
        Self::new(
            window.window_minutes,
            window.scope.as_deref(),
            window.used_percent,
            window.resets_at_unix,
        )
    }
}

/// An event the sidecar writes to flycod.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SidecarEvent {
    /// The Agent SDK imported successfully and the sidecar is listening.
    /// Always the first line, exactly once.
    Ready {
        /// Version of the installed `@anthropic-ai/claude-agent-sdk`.
        sdk_version: String,
    },
    /// The session is identified and its CLI is warming.
    ///
    /// Emitted as soon as the SDK query is constructed — before any user
    /// message — so the control plane can record the session and the UI can
    /// go live without waiting for a turn. The id is the sidecar's own
    /// (`Options.sessionId`) on a fresh session, and the resumed id
    /// otherwise.
    Started {
        /// Harness-native session id, used to resume this session later.
        session_id: String,
    },
    /// Capability tokens advertised by this CLI build.
    ///
    /// Feature detection reads this list and never a version string. It is
    /// its own event because the Agent SDK only reports it on the
    /// `system/init` stream frame, which the CLI emits at the start of a
    /// turn — so it necessarily arrives after
    /// [`SidecarEvent::Started`]. Emitted on every init frame; the newest
    /// set wins.
    Capabilities {
        /// The capability tokens, as the CLI names them.
        capabilities: Vec<String>,
    },
    /// Every model this CLI build offers, from the SDK's
    /// `supportedModels()`.
    ///
    /// Emitted once, as soon as the CLI has answered its `initialize`
    /// handshake and before any turn — the earliest moment the answer
    /// exists, and early enough that the composer's picker is right for the
    /// first message. Translated into flyco's own vocabulary in the
    /// sidecar rather than passed through raw, because the same shape has
    /// to come back from Codex.
    Models {
        /// The models, in the order the SDK listed them.
        models: Vec<flyco_core::ModelOption>,
    },
    /// How much of the account's plan is spent, from the SDK's
    /// `usage_EXPERIMENTAL_MAY_CHANGE_DO_NOT_RELY_ON_THIS_API_YET()`.
    ///
    /// Emitted once the CLI has answered its handshake, and again after
    /// every `result` message — the two moments the number can have moved.
    /// The list is empty for a session with no plan behind it at all (an
    /// API key, Bedrock, Vertex), which the SDK states outright with
    /// `rate_limits_available: false`.
    PlanUsage {
        /// Every window the SDK reported, in no particular order.
        windows: Vec<SidecarUsageWindow>,
    },
    /// Every slash command this session offers, from the SDK's
    /// `supportedCommands()`.
    ///
    /// Emitted once the CLI has answered its `initialize` handshake, and
    /// again on every `system/commands_changed` frame — the CLI discovers
    /// skills as the agent walks into subdirectories, so the set a session
    /// opens with is not the set it ends with. The newest list replaces the
    /// last one entirely, which is what the SDK documents that frame to
    /// mean.
    Commands {
        /// The commands, in the order the SDK listed them.
        commands: Vec<flyco_core::HarnessCommand>,
    },
    /// What the CLI actually mounted, from the SDK's `mcpServerStatus()`.
    ///
    /// Emitted once, as soon as the CLI has finished its `initialize`
    /// handshake and before any turn — the earliest moment the answer
    /// exists. flycod checks it against [`crate::mount::verify`] and fails
    /// the session if flyco's own server is not there, because an agent
    /// that cannot read its budget will spend past it.
    McpServers {
        /// One entry per server the CLI was told to mount.
        servers: Vec<MountedServer>,
    },
    /// One SDK stream message, verbatim. Interpreted in
    /// [`super::normalize`].
    SdkMessage {
        /// The raw SDK message.
        message: Value,
    },
    /// A `canUseTool` callback is blocked waiting for a decision.
    ApprovalRequest {
        /// Echo this in [`SidecarCommand::ApprovalDecision`].
        id: ApprovalId,
        /// Tool name the model asked to run.
        tool: String,
        /// Tool input as the model produced it.
        input: Value,
        /// The SDK's own suggested permission updates, when it offers any.
        suggestions: Option<Value>,
    },
    /// A `SessionStore` adapter method is blocked waiting for a result.
    StoreRequest {
        /// Echo this in [`SidecarCommand::StoreResponse`].
        id: StoreRequestId,
        /// What the store must do.
        op: StoreOp,
    },
    /// The sidecar cannot continue. Terminal: it exits after this line.
    Fatal {
        /// Human-readable cause.
        error: String,
    },
}

impl SidecarEvent {
    /// The `type` tag this event serializes under.
    #[must_use]
    pub const fn tag(&self) -> &'static str {
        match self {
            Self::Ready { .. } => "ready",
            Self::Started { .. } => "started",
            Self::Capabilities { .. } => "capabilities",
            Self::Models { .. } => "models",
            Self::PlanUsage { .. } => "plan_usage",
            Self::Commands { .. } => "commands",
            Self::McpServers { .. } => "mcp_servers",
            Self::SdkMessage { .. } => "sdk_message",
            Self::ApprovalRequest { .. } => "approval_request",
            Self::StoreRequest { .. } => "store_request",
            Self::Fatal { .. } => "fatal",
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{PermissionMode, SessionKey, SidecarAuth, SidecarCommand, SidecarEvent};

    #[test]
    fn unit_commands_are_bare_tags() {
        assert_eq!(
            serde_json::to_string(&SidecarCommand::Interrupt).expect("serialize"),
            r#"{"type":"interrupt"}"#
        );
        assert_eq!(
            serde_json::to_string(&SidecarCommand::Shutdown).expect("serialize"),
            r#"{"type":"shutdown"}"#
        );
        assert_eq!(
            serde_json::to_string(&SidecarCommand::Compact).expect("serialize"),
            r#"{"type":"compact"}"#
        );
    }

    #[test]
    fn permission_modes_use_the_sdks_camel_case() {
        for (mode, token) in [
            (PermissionMode::Default, "\"default\""),
            (PermissionMode::AcceptEdits, "\"acceptEdits\""),
            (PermissionMode::BypassPermissions, "\"bypassPermissions\""),
            (PermissionMode::Plan, "\"plan\""),
            (PermissionMode::DontAsk, "\"dontAsk\""),
            (PermissionMode::Auto, "\"auto\""),
        ] {
            assert_eq!(serde_json::to_string(&mode).expect("serialize"), token);
        }
    }

    #[test]
    fn an_absent_subpath_is_omitted_rather_than_null() {
        let key = SessionKey {
            project_key: "flyco".to_owned(),
            session_id: "s-1".to_owned(),
            subpath: None,
        };
        let json = serde_json::to_string(&key).expect("serialize");
        assert_eq!(json, r#"{"project_key":"flyco","session_id":"s-1"}"#);
    }

    #[test]
    fn auth_modes_are_tagged_on_mode() {
        let json = serde_json::to_value(SidecarAuth::ApiKey {
            key: "sk-test".to_owned(),
        })
        .expect("serialize");
        assert_eq!(json["mode"], "api_key");
        assert_eq!(json["key"], "sk-test");
    }

    #[test]
    fn capabilities_travel_separately_from_session_identity() {
        // The two cannot share an event: the SDK reports a session's
        // identity at construction and its capabilities only on the first
        // turn's `system/init`.
        let started = SidecarEvent::Started {
            session_id: "9d0f4b1a".to_owned(),
        };
        let json = serde_json::to_value(&started).expect("serialize");
        assert!(json.get("capabilities").is_none());

        let capabilities = SidecarEvent::Capabilities {
            capabilities: vec!["interrupt_receipt_v1".to_owned()],
        };
        let json = serde_json::to_value(&capabilities).expect("serialize");
        assert_eq!(json["type"], "capabilities");
    }

    #[test]
    fn tags_match_the_serialized_type_field() {
        let event = SidecarEvent::Ready {
            sdk_version: "0.3.250".to_owned(),
        };
        let json = serde_json::to_value(&event).expect("serialize");
        assert_eq!(json["type"], event.tag());
    }
}
